aws/resource_aws_lambda_function.go

```diff
@@ -685,101 +689,87 @@ func resourceAwsLambdaFunctionUpdate(d *schema.ResourceData, meta interface{}) e
+				TargetArn: aws.String(""),
+			}
+			if len(dlcMaps) == 1 { // Schema guarantees either 0 or 1
+				dlcMap := dlcMaps[0].(map[string]interface{})
+				configReq.DeadLetterConfig.TargetArn = aws.String(dlcMap["target_arn"].(string))
 			}
-			configUpdate = true
 		}
-	}
-	if d.HasChange("vpc_config") {
-		configReq.VpcConfig = &lambda.VpcConfig{
-			SecurityGroupIds: []*string{},
-			SubnetIds:        []*string{},
+		if d.HasChange("tracing_config") {
+			tracingConfig := d.Get("tracing_config").([]interface{})
+			if len(tracingConfig) == 1 { // Schema guarantees either 0 or 1
+				config := tracingConfig[0].(map[string]interface{})
+				configReq.TracingConfig = &lambda.TracingConfig{
+					Mode: aws.String(config["mode"].(string)),
+				}
+			}
 		}
-		if v, ok := d.GetOk("vpc_config"); ok && len(v.([]interface{})) > 0 {
-			vpcConfig := v.([]interface{})[0].(map[string]interface{})
-			configReq.VpcConfig.SecurityGroupIds = expandStringSet(vpcConfig["security_group_ids"].(*schema.Set))
-			configReq.VpcConfig.SubnetIds = expandStringSet(vpcConfig["subnet_ids"].(*schema.Set))
+		if d.HasChange("vpc_config") {
+			configReq.VpcConfig = &lambda.VpcConfig{
+				SecurityGroupIds: []*string{},
+				SubnetIds:        []*string{},
+			}
+			if v, ok := d.GetOk("vpc_config"); ok && len(v.([]interface{})) > 0 {
+				vpcConfig := v.([]interface{})[0].(map[string]interface{})
+				configReq.VpcConfig.SecurityGroupIds = expandStringSet(vpcConfig["security_group_ids"].(*schema.Set))
+				configReq.VpcConfig.SubnetIds = expandStringSet(vpcConfig["subnet_ids"].(*schema.Set))
+			}
 		}
-		configUpdate = true
-	}
 
-	if d.HasChange("runtime") {
-		configReq.Runtime = aws.String(d.Get("runtime").(string))
-		configUpdate = true
-	}
-	if d.HasChange("environment") {
-		if v, ok := d.GetOk("environment"); ok {
-			environments := v.([]interface{})
-			environment, ok := environments[0].(map[string]interface{})
-			if !ok {
-				return errors.New("At least one field is expected inside environment")
-			}
+		if d.HasChange("runtime") {
+			configReq.Runtime = aws.String(d.Get("runtime").(string))
+		}
+		if d.HasChange("environment") {
+			if v, ok := d.GetOk("environment"); ok {
+				environments := v.([]interface{})
+				environment, ok := environments[0].(map[string]interface{})
+				if !ok {
+					return errors.New("At least one field is expected inside environment")
+				}
 
-			if environmentVariables, ok := environment["variables"]; ok {
-				variables := readEnvironmentVariables(environmentVariables.(map[string]interface{}))
+				if environmentVariables, ok := environment["variables"]; ok {
+					variables := readEnvironmentVariables(environmentVariables.(map[string]interface{}))
 
+					configReq.Environment = &lambda.Environment{
+						Variables: aws.StringMap(variables),
+					}
+				}
+			} else {
 				configReq.Environment = &lambda.Environment{
-					Variables: aws.StringMap(variables),
+					Variables: aws.StringMap(map[string]string{}),
```

Comment:
This interesting diff is actually just me indenting the shown lines to nest within the conditional block.